design/healcheck-endpoint.md

```diff
@@ -0,0 +1,169 @@
+provide a class with similar capabilities that enables us to assemble a centralized Health reporting instance manually.
+
+Individual Healthcheck implementations, such as database check, External IdP check or LDAP check need to be provided
+as `HealthCheck` class implementations. Later on, they will be added to `KeycloakHealthReporter`.
+
+#### Keycloak REST integration
+
+Integrating with Wildfly's `/health` endpoint requires either adding CDI annotations to Keycloak and relying on CDI
+deployment processor (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java))
+or integrating with Microprofile Health Subsystem. An additional problem is that Wildfly provides its own
+`HealthReporter` class (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporter.java))
+that doesn't use groups. The Grouping functionality is essential for this design (explained below).
+
+The biggest downside of providing Keycloak's own implementation is that we'll end up with 2 health endpoints -
+one for Wildfly (`http://localhost:9990/health`) and one for Keycloak (`https://localhost:8443/auth/health`). We
+need to be very clear in the documentation that the latter is the proper one.
+
+Having the above in mind, the design assumes integrating the new Healthcheck API with Keycloak's REST interface.
+The implementation will create additional `/health` endpoint that will parse incoming `timeout` query parameter.
+
+#### Database Healthcheck
+
+One of the most important checks for the Healthcheck API is checking if the database is up. The Check is performed
+in the Passive Mode, meaning it analyzing metrics emitted by JCA subsystem. Currently, the JCA exposes the following metrics:
+- `wildfly_datasources_pool_in_use_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_idle_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_max_wait_time{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_timed_out{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_created_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+The implementation should include some experimentation, but the above 5 metrics should be enough to check the proper
+JCA state. The other metrics are available in [this document](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit?usp=sharing).
+Another aspect is that JCA subsystem needs to actively probe the data source.
+This can be achieved by configuring `check-valid-connection-sql` and `background-validation`
+(see the [Wildfly Model manual](https://docs.wildfly.org/19/wildscribe/subsystem/datasources/data-source/ExampleDS/index.html)).
```

Comment:
This section should cover thoughts on how to do "active" monitoring of database by having a min pool size set to 1, with a background check on connections

Reply:
I agree. Let me add it.